### PR TITLE
feat(java): add LanceOperation, commitAppend for batch write

### PR DESCRIPTION
**Pull request by LuQQiu**
Add LanceOperation, Add commit for Append Operation.
Pass Fragment as JSON string between Rust/Java.

### CODE REVIEW

File: java/core/lance-jni/src/blocking_dataset.rs [hunk 1]
```diff
@@ -42,6 +43,11 @@ impl BlockingDataset {
         Ok(Self { inner })
     }
 
+    pub fn commit(uri: &str, operation: Operation, read_version: Option<u64>) -> Result<Self> {
+        let inner = RT.block_on(Dataset::commit(uri, operation, read_version, None, None))?;
```
Comment (LuQQiu) [inline java/core/lance-jni/src/blocking_dataset.rs:47]: TODO commit without read_version will throw Unknown error which is hard for downstream to debug

**Comment by LuQQiu** [comment]
@eddyxu @beinan PTAL, thanks!

File: java/core/lance-jni/src/ffi.rs [hunk 2]
```diff
@@ -34,12 +37,27 @@ pub trait JNIEnvExt {
     /// Get Option<i64> from Java Optional<Long>.
     fn get_long_opt(&mut self, obj: &JObject) -> Result<Option<i64>>;
 
+    /// Get Option<u64> from Java Optional<Long>.
+    fn get_long_opt_u64(&mut self, obj: &JObject) -> Result<Option<u64>>;
```
Comment (eddyxu) [inline java/core/lance-jni/src/ffi.rs:41]: naming: `get_u64_opt`,

File: java/core/src/main/java/com/lancedb/lance/FragmentOperation.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+      validateFragments(updatedFragments);
+      validateFragments(deletedFragments);
+      this.updatedFragments = updatedFragments;
+      this.deletedFragments = deletedFragments;
+      this.predicate = predicate;
+    }
+
+    @Override
+    public Dataset commit(BufferAllocator allocator, String path, Optional<Integer> readVersion) {
+      throw new UnsupportedOperationException();
```
Comment (eddyxu) [inline java/core/src/main/java/com/lancedb/lance/FragmentOperation.java:74]: Do we need to implement this later?

File: java/core/src/main/java/com/lancedb/lance/FragmentMetadata.java [hunk 4]
```diff
@@ -14,15 +14,30 @@
...
-/** Metadata of a Fragment in the dataset. */
+import org.json.JSONObject;
+
+/**
+ * Metadata of a Fragment in the dataset. 
+ * Matching to lance Fragment.
+ * */
 public class FragmentMetadata {
-  private final int fragementId;
+  private final String jsonMetadata;
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply java/core/src/main/java/com/lancedb/lance/FragmentMetadata.java:24]: Keep the original json string to be able to pass it back to rust lance.
Yeah, make sense. Original plan to do one to one match and figure out there are many fields in the metadata and many of them are not needed by end-user. 
Will take out the needed fields and do the validation for those fields

File: java/core/src/main/java/com/lancedb/lance/FragmentOperation.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.vector.types.pojo.Schema;
+
+/** Fragment related operations. */
+public abstract class FragmentOperation {
+  protected static void validateFragments(List<FragmentMetadata> fragments) {
+    if (fragments == null || fragments.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply java/core/src/main/java/com/lancedb/lance/FragmentOperation.java:26]: Put the individual fragmentMetadata validation logics into its create function
FragmentMetadata.fromJson(String jsonMetadata)

File: java/core/lance-jni/src/ffi.rs (hunk 2 above)
Comment (LuQQiu) [reply java/core/lance-jni/src/ffi.rs]: Done

File: java/core/src/main/java/com/lancedb/lance/Fragment.java [hunk 6]
```diff
@@ -54,39 +38,19 @@ public static FragmentMetadata create(String datasetUri, BufferAllocator allocat
...
-      Optional<Integer> fragementId, WriteParams params) {
-    return new FragmentMetadata(createWithFfiStream(datasetUri, stream.memoryAddress(), fragementId,
+      Optional<Integer> fragmentId, WriteParams params) {
+    return new FragmentMetadata(createWithFfiStream(datasetUri, stream.memoryAddress(), fragmentId,
         params.getMaxRowsPerFile(), params.getMaxRowsPerGroup(),
         params.getMaxBytesPerFile(), params.getMode()));
   }
 
-  private static native int createWithFfiArray(String datasetUri,
+  private static native String createWithFfiArray(String datasetUri,
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply java/core/src/main/java/com/lancedb/lance/Fragment.java:53]: Added

File: java/core/src/main/java/com/lancedb/lance/FragmentOperation.java (hunk 3 above)
Comment (LuQQiu) [reply java/core/src/main/java/com/lancedb/lance/FragmentOperation.java]: In the current plan
1. Spark batch write + better arrow type & spark type conversion
2. Spark streaming write
3. may be partition support
May not be able to implement in the short time, so i delete them for now

**Comment by LuQQiu** [comment]
@eddyxu Addressed the comments, PTAL

**Review by eddyxu — APPROVED** [review]

**Pull request merged**